mysql-persistence/src/main/java/com/netflix/conductor/dao/mysql/MySQLExecutionDAO.java

```diff
@@ -0,0 +1,761 @@
+					continue;
+				}
+
+				insertOrUpdateTaskData(connection, task, INSERT_TASK);
+				addWorkflowToTaskMapping(connection, task);
+				addTaskInProgress(connection, task);
+				updateTask(connection, task);
+
+				created.add(task);
+			}
+		});
+
+		return created;
+	}
+
+	@Override
+	public void updateTask(Task task) {
+		withTransaction(connection -> updateTask(connection, task));
+	}
+
+	@Override
+	public boolean exceedsInProgressLimit(Task task) {
+		TaskDef taskDef = metadata.getTaskDef(task.getTaskDefName());
+		if (taskDef == null) return false;
+
+		int limit = taskDef.concurrencyLimit();
+		if (limit <= 0) return false;
+
+		long current = getInProgressTaskCount(task.getTaskDefName());
+
+		if (current >= limit) {
+			Monitors.recordTaskRateLimited(task.getTaskDefName(), limit);
+			return true;
+		}
+
+		logger.info("Task execution count for {}: limit={}, current={}", task.getTaskDefName(), limit, getInProgressTaskCount(task.getTaskDefName()));
+
+		String taskId = task.getTaskId();
+
+		List<String> tasksInProgressInOrderOfArrival = findAllTasksInProgressInOrderOfArrival(task, limit);
+
+		boolean rateLimited = !tasksInProgressInOrderOfArrival.contains(taskId);
+
+		if (rateLimited) {
+			logger.info("Task execution count limited. {}, limit {}, current {}", task.getTaskDefName(), limit, getInProgressTaskCount(task.getTaskDefName()));
+			Monitors.recordTaskRateLimited(task.getTaskDefName(), limit);
+		}
+
+		return rateLimited;
+	}
+
+	@Override
+	public void updateTasks(List<Task> tasks) {
+		withTransaction(connection -> tasks.forEach(task -> updateTask(connection, task)));
+	}
+
+	@Override
+	public void addTaskExecLog(List<TaskExecLog> log) {
+		indexer.add(log);
+	}
+
+	@Override
+	public void removeTask(String taskId) {
+		Task task = getTask(taskId);
+
+		if(task == null) {
+			logger.warn("No such Task by id {}", taskId);
+			return;
+		}
+
+		String taskKey = task.getReferenceTaskName() + "" + task.getRetryCount();
+
+		withTransaction(connection -> {
+			removeScheduledTask(connection, task, taskKey);
+			removeWorkflowToTaskMapping(connection, task);
+			removeTaskInProgress(connection, task);
+			removeTaskData(connection, task);
+		});
+	}
+
+	@Override
+	public Task getTask(String taskId) {
+		String taskJsonStr = getWithTransaction(c -> c.createQuery(GET_TASK).addParameter("taskId", taskId).executeScalar(String.class));
+		return taskJsonStr != null ? readValue(taskJsonStr, Task.class) : null;
+	}
+
+	@Override
+	public List<Task> getTasks(List<String> taskIds) {
+		if (taskIds.isEmpty()) return Lists.newArrayList();
+
+		List<String> taskIdParameters = taskIds.stream().map(taskId -> format("'%s'", taskId)).collect(Collectors.toList());
+
+		return getWithTransaction(c -> c.createQuery(format(GET_TASKS, join(taskIdParameters, ","))).executeScalarList(String.class))
+				.stream()
+				.filter(Objects::nonNull)
+				.map(taskJsonStr -> readValue(taskJsonStr, Task.class))
+				.collect(Collectors.toList());
```

Comment:
Elsewhere I've noticed that the result of this method is sorted. It would be nice to document that the return type must be mutable *and* avoid the `toList()` method which doesn't guarantee mutability.